[export] support non-persistent buffers

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118487
* #118486
* #118485
* #118484
* #118505
* #118483

Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

suo committed:
[export] support non-persistent buffers

Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

[ghstack-poisoned]

zhxchen17 commented:
> So I introduced some sad rewrite passes in _export make the final state dict correctly align with the original module's state dict.

now we are talking lol. 

joke aside, I feel making contract correct is more urgent than fixing aot autograd and dynamo. I'm fine this sort of rewriting as long as we have justified reason to make the contract aligned with eager.

> I can try that if people are interested in it.

Yes! I'm interested in that. I feel `persistent` is more like an attribute than an actual type of weight we need to keep track of everywhere (although making them as separate kind of input do force us to think more about how to implement them correctly), i.e. I'm not very convinced we should add an entirely new type of input in the graph signature level, since it mostly affects state_dict behavior, not much on other subsystems (e.g. functionalization, autograd, etc.)

suo commented:
Sounds good, let me try it that way. It actually aligns better with how we're serializing these values as well.

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

[ghstack-poisoned]

thiagocrepaldi reviewed (commented):
@titaiwangms FYI

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

suo committed:
Update on "[export] support non-persistent buffers"


Basic support for non-persistent buffers, which are buffers that do not show up in the state dict.

Basically, I added a new case to everything: there's a new `InputKind`, there's a new `InputSpec`, etc.

If we want to thicken the `InputKind` abstraction from just an enum to something that can hold some state, it might look nicer (you could check like `spec.kind.persistent` instead of checking whether `spec.kind` is `BUFFER` or `BUFFER_NON_PERSISTENT`). I can try that if people are interested in it.

The other weird twist is that most of our other systems (FX, aot_export, dynamo) have completely buggy handling of non-persistent buffers. I tried to go on a wild goose chase to fix them all, but it got to be too much. So I introduced some sad rewrite passes in `_export` make the final state dict correctly align with the original module's state dict.

This exposed some bugs/ambiguous handling of parameters/buffers in existing test code. For example, `TestSaveLoad.test_save_buffer` traced over a module that was not in the root module hierarchy and caused some weird behavior. I think we should error explicitly on use cases like this: https://github.com/pytorch/pytorch/issues/118410. For now I just rewrote the tests or skipped them.

Differential Revision: [D53011715](https://our.internmc.facebook.com/intern/diff/D53011715/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
